### DISCUSSION TITLE
Demo of setting a basic memory handler

### DISCUSSION THREAD

**Discussion by TysonRayJones** [NVIDIA/cuQuantum#82] [Q&A] [answered]
Hi there,

I am intending to bind cuStateVec to a mempool to avoid having to explicitly manage workspaces, as described in [this doc](https://docs.nvidia.com/cuda/cuquantum/latest/custatevec/api/functions.html#memory-management-api). I  wish to do this for code hygiene (of a "draft" cuQuantum implementation) rather than for performance, so I seek something simple. That doc shows the boilerplate necessary for using a custom memory pool in a very general way, but I am wondering whether this can be reduced/simplified if using a device's default pool.

Here is how the doc might suggest using the default pool:

```C++
// get the default mem pool (assuming single GPU)
int device = 0;
cudaMemPool_t memPool;
cudaDeviceGetMemPool(&memPool, device);

// optionally tweak it here (affecting existing pool), e.g.
cudaMemPoolSetAttribute (memPool, cudaMemPoolAttrReleaseThreshold, 16*(1LL<<10)); 

// make mem pool alloc & dealloc callbacks for mem handler
int my_alloc(void* ctx, void** ptr, size_t size, cudaStream_t stream) {
    cudaMemPool_t pool = * reinterpret_cast<cudaMemPool_t*>(ctx);
    return cudaMallocFromPoolAsync(ptr, size, pool, stream); 
}
int my_dealloc(void* ctx, void* ptr, size_t size, cudaStream_t stream) {
    return cudaFreeAsync(ptr, stream); 
}

// create a mem handler around the mem pool
custatevecDeviceMemHandler_t handler;
handler.ctx = reinterpret_cast<void*>(&memPool);
handler.device_alloc = my_alloc;
handler.device_free = my_dealloc;

// set mem handler to auto-manage workspaces (handle = custatevecHandle_t)
custatevecSetDeviceMemHandler(handle, &handler);
```

This seems unnecessarily tedious to me, given I'm not really specifying any custom behavior; just mapping `device_alloc` to `cudaMallocFromPoolAsync` and `device_free` to `cudaFreeAsync`.
I sort of imagine that this setup (using the default pool) is what _should_ happen if one calls `custatevecApplyMatrix` (for example) with `extraWorkspace=nullptr`, without having previously called `custatevecSetDeviceMemHandler()`.

Is above the *right way* to use the default mem pool? Or is there a reason I should avoid using an existing pool?
If this is all fine, and I understand correctly this will be a common use-case (especially for new cuQuantum users not intimately familiar with CUDA), could there exist a bespoke function to avoid this boilerplate?
E.g.
```C++
custatevecSetDeviceMemHandlerToDefaultMemPool();
```
(and of course such a function would be expected to error if the user's device does not support stream-ordered memory)

In any case, it might be helpful to mention in the cuStateVec doc that the default mem pool (rather than a custom one) can be used.

**Comment by mtjrider** [comment]
@TysonRayJones this is a great question.
I've migrated this from an issue to a Q&A discussion.

We're confirming a few things, and will get back to you soon.

**Comment by ymagchi** [comment, marked as answer]
Hi @TysonRayJones, Thank you for your interest.

Your code above should work. `cudaDeviceGetDefaultMemPool` would be also available if you could call `cudaDeviceSetMemPool` before `cudaDeviceGetMemPool`. Our sample code can be found in [memory_handler.cu](https://github.com/NVIDIA/cuQuantum/blob/c202ee7ad23ce9a5072a32b45d36e8310a20eeb5/samples/custatevec/memory_handler.cu).

Currently we don't provide a function to attach `cudaMemPool_t`, as we'd like to allow flexibility to use other memory allocators (e.g., RAPIDS Memory Manager).

> In any case, it might be helpful to mention in the cuStateVec doc that the default mem pool (rather than a custom one) can be used.

Thanks for the suggestion. We will update that section in a future release.

**Reply by TysonRayJones** [reply]
Hi @ymagchi,

Ahh I understand - treating custom allocators as first-class citizens in the API is pretty admirable!
I'll use this chance to say that the cuStateVec API is (imo) beautifully designed - a set of few functions with thorough/comprehensive parameters.
The scoping of simulator facilities, and boiling them down to their essential core functions, was excellently done!

Thanks very much for the sample code, it's a great help.

**Reply by mtjrider** [reply]
Thanks for your feedback @TysonRayJones 
Memory pools and their implementation is a pervasive topic throughout the PyData ecosystem. We don't pretend to know how all of our users might want to integrate with cuQuantum.

To that end, if you have any additional insights or feedback you'd like to share regarding usage patterns, please let us know.

**Reply by TysonRayJones** [reply]
I'm far from a memory pool (or a CUDA) expert, but I'm sure I'll have lots of feedback about general usage patterns in time!

For the moment, I'm still struggling to get this "automatic workspace" to work. 
With my configuration above (and a `cuStream` bound to the `cuStateVec` handle), I cannot get functions like `custatevecApplyMatrix` (with a `nullptr` workspace and `0` work size) to succeed with `5` or more target qubits. This is naturally the threshold whereby non-zero workspace memory is _needed_ (`1 MiB` precisely). The call will instead strangely reset the vector to all zeroes.

While trying to prepare a MWE for this, I discovered I couldn't even get [that example](https://github.com/NVIDIA/cuQuantum/blob/c202ee7ad23ce9a5072a32b45d36e8310a20eeb5/samples/custatevec/memory_handler.cu) to run with states of `5` or more qubits. When I try to call `custatevecInitializeStateVector` upon a `5`-qubit state-vector that was asynch stream allocated (`cudaMallocAsync`), I receive error `Error: an illegal memory access was encountered`.
I'm attempting this on a Quadro RTX 5000 (`CC = 7.5`) and CUDA 12.2, and all checks of mem-pool compatibility are passing.

The MWE (which I simply compile with `nvcc -lcustatevec`):

```C++
#include <cuda_runtime_api.h>
#include <cuComplex.h>        // cuDoubleComplex
#include <custatevec.h>       // custatevecApplyMatrix
#include <string.h>           // strcpy
#include <stdio.h>            // printf
#include <stdlib.h>           // EXIT_FAILURE



#define HANDLE_ERROR(x)                                                        \
{   const auto err = x;                                                        \
    if (err != CUSTATEVEC_STATUS_SUCCESS ) {                                   \
        printf("Error: %s in line %d\n",                                       \
               custatevecGetErrorString(err), __LINE__); return err; }         \
};

#define HANDLE_CUDA_ERROR(x)                                                   \
{   const auto err = x;                                                        \
    if (err != cudaSuccess ) {                                                 \
        printf("Error: %s in line %d\n",                                       \
               cudaGetErrorString(err), __LINE__); return err; }               \
};


int myMemPoolAlloc(void* ctx, void** ptr, size_t size, cudaStream_t stream) {
    cudaMemPool_t& pool = *static_cast<cudaMemPool_t*>(ctx);
    cudaError_t status = cudaMallocFromPoolAsync(ptr, size, pool, stream);
    return (int)status;
}
int myMemPoolFree(void*, void* ptr, size_t, cudaStream_t stream) {
    cudaError_t status = cudaFreeAsync(ptr, stream);
    return (int)status;
}




int main(void) {

    printf("CUDA version: %d.%d\n", __CUDACC_VER_MAJOR__, __CUDACC_VER_MINOR__);

    // state vector
    const int nIndexBits   = 5;
    const int nSvSize      = (1 << nIndexBits);


    //----------------------------------------------------------------------------------------------
    // device configuration
    int deviceId;
    HANDLE_CUDA_ERROR( cudaGetDevice(&deviceId) );

    cudaError_t status;
    int isMemPoolSupported;
    status = cudaDeviceGetAttribute(&isMemPoolSupported, cudaDevAttrMemoryPoolsSupported, deviceId);
    if (status != cudaSuccess || !isMemPoolSupported) {
        printf("memory handler example WAIVED: CUDA Memory pools is not supported.\n");
        return EXIT_SUCCESS;
    }

    cudaMemPool_t memPool;
    HANDLE_CUDA_ERROR( cudaDeviceGetDefaultMemPool(&memPool, deviceId) );

    // avoid shrinking the pool (does not affect below failing of |+> init
    uint64_t threshold = UINT64_MAX;
    cudaMemPoolSetAttribute(memPool, cudaMemPoolAttrReleaseThreshold, &threshold);

    cudaStream_t stream;
    HANDLE_CUDA_ERROR( cudaStreamCreate(&stream) );


    //----------------------------------------------------------------------------------------------
    // creation of state vector
    cuDoubleComplex *d_sv;
    HANDLE_CUDA_ERROR( cudaMallocAsync((void**)&d_sv, nSvSize * sizeof(cuDoubleComplex), stream) );


    //----------------------------------------------------------------------------------------------
    // custatevec handle initialization
    custatevecHandle_t handle;
    HANDLE_ERROR( custatevecCreate(&handle) );
    HANDLE_ERROR( custatevecSetStream(handle, stream) );

    // device memory handler
    custatevecDeviceMemHandler_t handler;
    handler.ctx = &memPool;
    handler.device_alloc = myMemPoolAlloc;
    handler.device_free = myMemPoolFree;
    strcpy(handler.name, "mempool");
    HANDLE_ERROR( custatevecSetDeviceMemHandler(handle, &handler) );


    // ---------------------------------------------------------------------------------------------
    // when nIndexBits >=5, this state vector initialization fails with error:
    // Error: an illegal memory access was encountered

    cudaStreamSynchronize(stream);
    custatevecInitializeStateVector(
        handle, d_sv, CUDA_C_64F, nSvSize,
        CUSTATEVEC_STATE_VECTOR_TYPE_UNIFORM);


    //----------------------------------------------------------------------------------------------

    // clean up
    HANDLE_CUDA_ERROR( cudaStreamSynchronize(stream) );
    HANDLE_ERROR( custatevecDestroy(handle) );
    HANDLE_CUDA_ERROR( cudaFreeAsync(d_sv, stream) );
    HANDLE_CUDA_ERROR( cudaStreamDestroy(stream) );
}
```

It almost appears like the `cudaMallocAsync` (of the state-vector creation) is silently failing, due to some the issue with the memory pool, rather than with cuStateVec. _However_, when I try to directly modify the allocated memory in a custom kernel, it appears to work fine:
```C++
__global__ void test_set(cuDoubleComplex* vec, int size){
    long long int index;

    index = blockIdx.x*blockDim.x + threadIdx.x;
    if (index>=size) return;

    vec[index] = make_cuDoubleComplex(1,1);
}

\\ ...
\\ replace custatevecInitializeStateVector with:
int threadsPerCUDABlock = 128;
int CUDABlocks = ceil(size/(double) threadsPerCUDABlock);
test_set<<<CUDABlocks, threadsPerCUDABlock>>>(d_sv, size);
```

Without seeing how `custatevecInitializeStateVector` is modifying `d_sv`, I can't speculate any further. 
Any advice on what I should try next? I'm stumped!

**Reply by ymagchi** [reply]
For the code above, it seems that invalid memory accesses occurred by wrong parameters in `custatevecInitializeStateVector`.
Could you please try replacing `nSvSize` with `nIndexBits` as follows?
```
    custatevecInitializeStateVector(
        handle, d_sv, CUDA_C_64F, nIndexBits,
        CUSTATEVEC_STATE_VECTOR_TYPE_UNIFORM);
```
The descriptions for this API can be found in [cuStateVec Functions - custatevecInitializeStateVector](https://docs.nvidia.com/cuda/cuquantum/latest/custatevec/api/functions.html#custatevecinitializestatevector).

**Reply by TysonRayJones** [reply]
Oh woops, sorry for an embarrassing error! With that resolved, I _can_ get an isolated 5-qubit example to work with memory pools, but **_not_** my actual project (which potentially complicates matters with Thrust and Catch2).

 My memory setup between tests are identical:
```
create cuStateVec handle
create stream
bind stream to handle
obtain default mem pool
make mem handler
bound mem handler to handle
```
I'll report back for peace-of-mind when I've chased the bug further - I have one working copy where `custatevecApplyMatrix` seg-faults, and another where it leaves the state unmodified whenever `nTargets >= 5`. Both very likely my fault, though they _appear_ to differ from my working MWE only by their (elsewhere) integration of Thrust and Catch2!

In case you're curious, I'm debugging [this branch](https://github.com/QuEST-Kit/QuEST/tree/cuquantum-mempool-debug), and the cuStateVec setup is done [here](https://github.com/QuEST-Kit/QuEST/blob/c51e0039e64c72b42cada9571dba18a7e570953e/QuEST/src/GPU/QuEST_cuQuantum.cu#L196). 

Not that it's worth the time, but the quickest way to witness the error is to replace the empty `initDebugState()` in that file with
```C++
__global__ void debugKernel(cuAmp* vec, int size){
    long long int index = blockIdx.x*blockDim.x + threadIdx.x;
    if (index>=size) return;
    vec[index] = TO_CU_AMP(.2*index, .2*index+.1);
}

void statevec_initDebugState(Qureg qureg) {

    cudaStreamSynchronize(qureg.cuStream);

    int threadsPerCUDABlock = 128;
    int CUDABlocks = ceil(qureg.numAmpsTotal/(double) threadsPerCUDABlock);
    debugKernel<<<CUDABlocks, threadsPerCUDABlock>>>(qureg.deviceCuStateVec, qureg.numAmpsTotal);

    cudaStreamSynchronize(qureg.cuStream);
}
```
and run the failing unit test (on unix) with
```
export MY_COMPUTE_CAPABILITY=75

mkdir build
cd build

cmake .. -DTESTING=ON -DGPUACCELERATED=1 -DGPU_COMPUTE_CAPABILITY=$MY_COMPUTE_CAPABILITY -DUSE_CUQUANTUM=ON
make

./tests/tests multiQubitUnitary -a
```
where the `411`th test (the first of `5` qubits, and the first to require workspace memory) will fail.

Anyway I'll report back either to reveal my embarrassing error, or (significantly less likely) to raise a compatibility issue with Thrust + Catch2 :^)

**Reply by TysonRayJones** [reply]
Solved at last! I had obfuscated the problem by introducing other issues (giving stack structs to functions like `custatevecSetDeviceMemHandler()` which were cleared during execution).
 
The core issue was two missing `&` in my addressing of the fields of `custatevecDeviceMemHandler_t`, neither of which raised compiler issues due to `(void *)` typing, but miraclously neither seg-faults.

These were in the memory pool allocator callback:
```C++
int memPoolAlloc(void* ctx, void** ptr, size_t size, cudaStream_t stream) {
     // neglected & before pool
    cudaMemPool_t& pool = *static_cast<cudaMemPool_t*>(ctx);
    return cudaMallocFromPoolAsync(ptr, size, pool, stream); 
}
```
and in the binding of the mem handler's `ctx` field;
```C++
// below, myMemHandler is type myCustatevecDeviceMemHandler_t
// neglected & before myMemPool
myMemHandler.ctx = &myMemPool;
myMemHandler.device_alloc = memPoolAlloc;
// ...
```

Neglecting one or both of these `&` caused insidious errors, without compiler warnings nor crashes. Instead, cuStateVec would run and simply give erroneous results whenever an API function needed a non-zero workspace. 

I argue the mistake is easy to make, since the user will have assuredly more complicated code than [this example](https://github.com/NVIDIA/cuQuantum/blob/c202ee7ad23ce9a5072a32b45d36e8310a20eeb5/samples/custatevec/memory_handler.cu) wherein the `cudaMemPool_t`,  `cudaStream_t`, `custatevecHandle_t` and `custatevecDeviceMemHandler_t`were all in-stack in the top-level function. Instead, these types will be managed in existing library memory management (like in my example, [here](https://github.com/QuEST-Kit/QuEST/blob/f7c6ee6d8b07bc1fb9b6a95172e424761dc001fb/QuEST/src/GPU/QuEST_cuQuantum.cu#L147)), likely in the heap, and have more complicated addressing.

All in all, this is somewhat of a _dangerous_ mistake, and a very painful one to debug! I wonder if this can be detected somehow by `custatevecSetDeviceMemHandler()` at runtime, and returned as an error. 
Perhaps in the near term it's not unreasonable to caution in the documentation something like:
> Because `myCustatevecDeviceMemHandler_t` is agnostic to the form and type of its bound memory pool, some type and addressing errors cannot be detected at compile time and may lead to silent, erroneous workspace behaviour. For instance, `.ctx` must match the level of pointer redirection expected by function `.device_alloc`.

Perhaps it's too obvious for a general CUDA programmer, but could be a valuable warning to less experienced CUDA programmers (like myself) integrating cuStateVec `¯\_(ツ)_/¯ `

**Reply by mtjrider** [reply]
Thanks for your feedback, @TysonRayJones 

As you aptly pointed out, a memory pool must be described as a reference variable with its corresponding location in the handler.
Are you asking if we can confirm the validity of memory pool location provided to the handler in our library?

It's certainly true that we can improve our documentation on this subject. I've taken the action to do just that.
I'd like to clarify any requirements on validation w.r.t. our library APIs.
